### ISSUE TITLE
Undoing a rename that caused a file name to change doesn't revert the file name

### ISSUE THREAD

**Issue by JosephTLyons** [zed-industries/zed#5369]
[Undoing a rename that caused a file name to change doesn't revert the file name](https://user-images.githubusercontent.com/19867440/154830861-69cbb989-1a40-45e8-b0af-b0ca18bed85e.mov)

**Comment by petros** [comment]
Hey @JosephTLyons is this still an issue? I watched the video but I am not entirely sure how to reproduce this. Could you share some steps and what's expected instead? I think you use keyboard shortcuts in the video but I am not sure which ones.

**Comment by JosephTLyons** [comment]
Yeah for sure. I just went through and tested this again and it is still an issue on my end.

Here is approximately the same line of code as in the video:

- https://github.com/JosephTLyons/ascii_clock/blob/1f7f37c765a369017f6d3b6dc444e00f503a1fdd/src/clock.rs#L91

What I did was a `Rename Symbol` action on `clock_characters` - I renamed it `clock_characterssss`. Doing this also adjusts the name of the file that that symbol is associated with: `clock_characterssss.rs`.

I think the language server does this. Running an undo does revert the changes to the code in the current file (`clock.rs`), but it does not rename the `clock_characterssss.rs` file back to `clock_characters.rs`, which causes the code to go into a broken state.

I feel like this will take some custom logic of holding the previous file name in memory, but maybe I'm wrong and there's something offered by the language server that we aren't tapping into - so this could either be challenging or easy, depending.

**Comment by petros** [comment]
Thank you @JosephTLyons! I can reproduce it now. Indeed, I think I found the code that does this, but I am not entirely sure what the best approach is to include reverting the filename change with the undo action. I am going to investigate further.

**Comment by HactarCE** [comment]
I'm still able to reproduce this.

**Comment by ivanpasq** [comment]
I'm still able to reproduce the issue both in stable channel and in `main` branch. I'm trying to look into it.

**Comment by petros** [comment]
I can still reproduce this, and I am working on a PR (finally 😄) that I will open soon.

**Comment by petros** [comment]
Phew! This has been quite an adventure.

I had a go at this and got two working implementations, then abandoned both. Writing up what I found in case it saves someone the detour.

### What happens

`LspStore::deserialize_workspace_edit` applies `ResourceOp::Rename` straight to disk, then returns a `ProjectTransaction`, which only maps buffers to text transactions. The move is never recorded, so undo has nothing to reverse.

### First attempt

Record the moves on the transaction, keep a map from transaction id to moves in the editor, reverse them in `Editor::undo`.

Three things go wrong. Buffer and multibuffer transaction ids are both `clock::Lamport` drawn from unrelated clocks that both start at replica 0, so they collide and an unrelated undo can move files on disk. The map lives on the editor and `Editor::clone` does not copy it, so undoing in a split strands the file for good. And `Editor::undo` is only one of the undo entry points, since vim's `UndoLastLine`, the agent's edit rejection in `action_log` and `buffer_codegen` all call the buffer's undo directly.

### Second attempt

Key the moves by `(BufferId, TransactionId)` and emit `TransactionUndone` and `TransactionRedone` from `language::Buffer`. Every undo path bottoms out there, so this covers all of them, and the editor stops knowing about renames at all.

This one survives splits, covers vim and the agent, and works over SSH once the client registers what it decodes off the wire. It is still wrong, for a reason I do not think this layer can fix.

### The part I could not solve

A workspace edit is one user action but becomes N buffer transactions with N independent undo stacks. The move belongs to the action, and there is nowhere to put it.

Say a rename edits `a.ts` and `b.ts` and moves `a.ts` to `A.ts`. Undo in `a.ts` and the file moves back. Redo in `a.ts` and it moves forward. Now undo in `b.ts`. Whatever happens next is wrong, because "is this edit undone" has no answer while the two stacks disagree. Different key press orders leave the filesystem in different states for the same final text.

`text::History` is per buffer, so everything I tried was a proxy for the same missing thing, an undoable entry that belongs to the edit rather than to one buffer's history.

### Traps, if someone else tries

- Reversing with `Project::rename_entry` sends `willRenameFiles` and applies the edits it returns with `push_to_history: false`, so they land on top of the text undo has just restored and cannot themselves be undone. ~~`WorktreeStore::rename_entry` plus sending `didRenameFiles` yourself avoids it. The request is the trap, the notification is still needed.~~ **Correction:** that only holds on a local project. On a remote one `WorktreeStore::rename_entry` sends `proto::RenameProjectEntry`, and the only registered handler for it calls `will_rename_entry` on the host before moving anything, so the edits arrive regardless and there is no way to opt out. It also only bites when the reversal is automatic. If the user asked for the move, the server updating references is the same thing that happens for any rename from the project panel.
- `Fs::rename` returns `Ok(())` without moving anything when `ignoreIfExists` is set and the target already exists, so recording that as a move points undo at a file the refactor never touched.
- No fake server in the suite advertises `workspace.fileOperations`, so `will_rename_entry` and `did_rename_entry` both bail and CI never runs the branch production takes.
- The forward path saves the buffer before renaming, so even a perfect path reversal leaves the file holding content from the future. Whether undo should restore contents too is probably a separate decision.

### One unrelated fix

~Chasing the `ignoreIfExists` case above turned up something separate. `FakeFs::rename` does not match `RealFs`: where the real one reports success and leaves both files where they are, the fake removes the source. So a rename that should be a no-op quietly destroys a file, and no test can currently observe the correct behavior. It has nothing to do with this issue beyond being how I found it, so I will send it as its own small PR.~ Fixed in https://github.com/zed-industries/zed/pull/63416

### Where that leaves it

There are two paths from here, as far as I can think. A steer on which one is worth pursuing would be welcome before I spend more time on either.

#### Grouped undo

This is the one that looks correct to me. An undoable unit living above buffers would own both the buffer transactions an edit produced and its resource operations. `multi_buffer::Transaction` already has roughly that shape, but it only applies when undo runs through a multibuffer, and the direct `Buffer::undo` callers bypass it. So per buffer undo would have to become group aware, undoing a whole group or refusing. That changes what undo means rather than adding to it, which is why I am treating it as a prototype rather than a final PR.

#### Tell the user

The smaller option. Undo reverts the text, the user is told the file is still renamed, and an explicit action moves it back. That concedes the move is not part of the undo, but it puts a person in the loop instead of guessing which stack to believe, and the project stops being left quietly broken.

I am going to prototype Grouped undo and build Tell the user, then follow up here. If either looks wrong to someone who knows this code better, I would be glad to hear that before I get too far in.

#### My previous failed attempts

Both attempts are on my fork if the detail is useful. They are reference points rather than proposals, and each commit message records what I knew was broken at that stage.

- [attempt 1, editor owned](https://github.com/petros/zed/tree/5369-attempt-1-editor-owned)
- [attempt 2, buffer events](https://github.com/petros/zed/tree/5369-attempt-2-buffer-events)

**Comment by petros** [comment]
Follow-up on my last comment.

@devzeth described a second symptom above in Feb 2025, where the first `Rename symbol` moves the file but leaves the references in the other file alone, and @dybucc confirmed it. That turned out to be a separate bug from the undo one.

Zed tells servers it supports `workspace/willRenameFiles`. rust-analyzer takes that as a promise, so for a module rename it leaves the reference edits out of the `textDocument/rename` response on purpose and waits to be asked for them when the file moves. That is how it stops a client from applying them twice, see microsoft/vscode-languageserver-node#752. Zed does ask when you rename a file from the project panel, but not when a workspace edit moves one, so for a module rename nobody ever asks.

The [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_willRenameFiles) expects the request in both cases. It is sent "before files are actually renamed as long as the rename is triggered from within the client either by a user action or by applying a workspace edit".

I measured it by driving rust-analyzer straight over LSP with Zed's client capabilities, which takes Zed out of the picture:

- the rename response carries one rename operation and no text edits
- drop `fileOperations` from the client capabilities and all three reference edits arrive
- keep it and send `workspace/willRenameFiles`, and the same three edits arrive

Builds from March, July and August behave the same, so this is not a regression. Renaming a plain function is fine, so it is module renames specifically.

#63517 asks, and folds the answer into the transaction the edit is building so that undo reaches those edits. It also sends `didRenameFiles` afterwards, which this path never did.

## The other two

#63481. A workspace edit that asks for a directory to be deleted always fails, so whatever the edit did before that point is left half applied. I hit it building a Go repro for the toast below. It is not Go specific, any server sending a directory delete runs into it.

#63483. The "Tell the user" option from my last comment. Undo reverts the text, a toast reports that the file is still renamed, and an action moves it back.

The three are independent and none of them depends on the others. Together, a Rust module rename updates its references, and undoing it tells you the file is still moved and offers to move it back.

**Comment by humdrum00001010** [comment]
@petros workspace rename then undo fails too. I opened that issue, but it seems that it belongs here, close
